sfe/sfe.go

```diff
@@ -0,0 +1,287 @@
+		// submitting the form.
+		renderTemplate(response, "unpause-params.html", tmplData{unpausePath, string(incomingJWT)})
+	} else {
+		// We only want to accept requests containing the JWT param.
+		renderTemplate(response, "unpause-noParams.html", nil)
+	}
+}
+
+// posthelper After clicking unpause, serve a page indicating if the unpause succeeded or failed.
+func (sfe *SelfServiceFrontEndImpl) postHelper(response http.ResponseWriter, incomingJWT unpauseJWT) {
+	if incomingJWT != "" {
+		type tmplData struct {
+			ShouldUnpause bool
+			AccountID     string
+		}
+
+		claims, err := sfe.validateJWTforAccount(incomingJWT)
+		if err != nil {
+			renderTemplate(response, "unpause-post.html", nil)
+		}
+
+		// TODO(#7356) Declare a registration ID variable to populate an
+		// rapb unpause account request message.
+		_, innerErr := strconv.ParseInt(claims.Subject, 10, 64)
+		if innerErr != nil {
+			renderTemplate(response, "unpause-noParams.html", nil)
+		}
+
+		// TODO(#7536) Send gRPC request to the RA informing it to unpause
+		// the account specified in the claim. At this point we should wait
+		// for the RA to process the request before returning to the client,
+		// just in case the request fails.
+
+		// Success, the account has been unpaused.
+		renderTemplate(response, "unpause-post.html", tmplData{true, claims.Subject})
+	} else {
+		renderTemplate(response, "unpause-noParams.html", nil)
+	}
+}
+
+// Unpause allows a requester to unpause their account via a form present on the
+// page.
+func (sfe *SelfServiceFrontEndImpl) Unpause(response http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodHead:
+		return
+	case http.MethodGet:
+		sfe.getHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	case http.MethodPost:
+		sfe.postHelper(response, unpauseJWT(request.URL.Query().Get("jwt")))
+	default:
+		response.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, POST")
+		response.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+}
+
+type sfeJWTClaims struct {
+	jwt.Claims
+
+	// Version is a custom claim used to mass invalidate existing JWTs by
+	// changing the API version via unpausePath.
+	Version string `json:"apiVersion,omitempty"`
+}
+
+// validateJWT derives a ed25519 public key from a seed shared by the SFE and
+// WFE. The public key is used to validate the signature and contents of an
+// unpauseJWT and verify that the its claims match a set of expected claims.
+// Passing validations returns the claims or an error.
+func (sfe *SelfServiceFrontEndImpl) validateJWTforAccount(incomingJWT unpauseJWT) (sfeJWTClaims, error) {
+	slug := strings.Split(unpausePath, "/")
+	if len(slug) != 4 {
+		return sfeJWTClaims{}, errors.New("Could not parse API version")
+	}
+
+	token, err := jwt.ParseSigned(string(incomingJWT), []jose.SignatureAlgorithm{jose.EdDSA})
+	if err != nil {
+		return sfeJWTClaims{}, fmt.Errorf("parsing JWT: %s", err)
+	}
+
+	incomingClaims := sfeJWTClaims{}
```

Comment:
Like `makeJWTForAccount`, this should be moved to a "pause" directory. It should be exported for use in the SFE business logic and WFE tests.

Reply:
Once this is moved, I suppose all we really need to do is rename this `pause.ValidateJWT()`. This endpoint isn't authenticated so we don't really even know who the requester actually is and we don't care who they purport to be (e.g., we wouldn't want to validate this against a query param). Possession of a valid JWT is all we care about and the RegID can be found there-in.